Added exclude_from_copy to config

KabaevRoman opened the pull request:
## Description

Fixes #916.

<!-- Description of the changes introduced by this PR. -->
Added exclude_from_copy to config, to exclude some paths via globs
## TODOs

Read the [Gruntwork contribution guidelines](https://gruntwork.notion.site/Gruntwork-Coding-Methodology-02fdcd6e4b004e818553684760bf691e).

- [x] Update the docs.
- [x] Run the relevant tests successfully, including pre-commit checks.
- [x] Ensure any 3rd party code adheres with our [license policy](https://www.notion.so/gruntwork/Gruntwork-licenses-and-open-source-usage-policy-f7dece1f780341c7b69c1763f22b1378) or delete this line if its not applicable.
- [ ] Include release notes. If this PR is backward incompatible, include a migration guide.

## Release Notes (draft)

<!-- One-line description of the PR that can be included in the final release notes. -->
Added / Removed / Updated [X].

### Migration Guide

<!-- Important: If you made any backward incompatible changes, then you must write a migration guide! -->

KabaevRoman commented:
Don't know if exclude_in_copy syntactically correct, if I should rename exclude_from_copy to exclude_in_copy, or made mistakes in formatting/code let me know!

KabaevRoman commented:
config tests > https://gist.github.com/KabaevRoman/17cbee1e7cacb6e154d0ae18945c898f 
util tests > https://gist.github.com/KabaevRoman/f7f4bc1aa456d9f0326547db8fa4b4bf

yhakbar reviewed (changes requested):
Hey @KabaevRoman,

Sorry we haven't gotten around to reviewing this until now. I'll look to kick off CI later today.

In the meantime, do you mind adding documentation and testing specifically around the scenario where files are listed as both included _and_ excluded? I would assume that the safest thing to do would be to exclude them in that circumstance.

You should be able to run most of the tests in this repository without an AWS account, etc. We've done some work fairly recently to move all tests that require external integrations to require build tags, so they shouldn't run by default.

You might want to be targeted with your tests by doing something like this anyways, though:

```bash
go test -run 'NameOfTest'
```

We have a goal of making it so that contributors can easily run tests locally, so any help you could provide to make that a better experience would be much appreciated.

KabaevRoman commented:
Hi there, missed previous message,
I've updated the docs here https://github.com/gruntwork-io/terragrunt/pull/3543/files#diff-eac47411aff1ec957598f585714c7199a1c1c31a710ef472868bae5abef38cf9R111, do i need to update them somewhere else?
As for tests I've added tests for behaviour that you described here https://github.com/gruntwork-io/terragrunt/pull/3543/files#diff-4146922e3bcd164a817418c48bfd4ee62b79863281d449038daab9109c81cfbdR431, i think you can re-open pr and i will resolve existing conflicts.
Give me a hint what tests should i run for this pr, currently my tests are ok, but I'm a bit worried that I could break something,
but running all tests locally as described in docs seems to be not working properly, cause I get errors with some .hcl test files paths or something like that.

yhakbar commented:
No problem, @KabaevRoman ! I'll re-open the PR as you're up for picking it up.

The documentation I'm looking to see here is very explicitly telling users what happens when both `include_in_copy` and `exclude_from_copy` are defined.

I see that you added this:

```
  When `include_in_copy` is provided, you can still provide `exclude_from_copy` to skip provided glob, or nested glob from `include_in_copy` 
```

But as a user reading these docs, I would rather see something like this:

```
Note that using `include_in_copy` and `exclude_from_copy` are not mutually exclusive. 

If a file matches a pattern in both `include_in_copy` and `exclude_from_copy`, it will not be included. If you would like to ensure that the file _is_ included, make sure the patterns you use for `include_in_copy` do not match the patterns in `exclude_from_copy`.
```

You don't need to run all the tests. Only run the tests that you think are relevant locally. Once you've resolved conflicts, I can re-run tests for you in CI. Note that quite a lot of work is currently underway for the code paths that you're touching here https://github.com/gruntwork-io/terragrunt/pull/3723, so we might need to wait a little bit to merge this so that we can keep development unblocked for Terragrunt 1.0 work.

KabaevRoman commented:
I've updated pr and changed readme as you advised, tests seems to be passing

yhakbar commented:
Thanks @KabaevRoman . I've kicked off CI. I'll coordinate with the team to see if we can merge this in earlier at the risk of increased conflicts, or if we should hold off for a bit. Either way, thanks for your patience and working with us to get this done!

KabaevRoman commented:
Some tests failed in pipeline, could discover linting errors and how to run linters from makefile and fixed them, but got no luck with running integration tests properly even unit_test from config.yml. Also i have no access to pipeline logs to get more details for failing tests. And have no idea what is this shortcut you are using `run-go-tests`

KabaevRoman commented:
Fixed test that you mentioned, it passes now. Maybe i will describe later all issues i encountered, on running tests linters etc. It seems like docs are a bit outdated https://terragrunt.gruntwork.io/docs/community/contributing/#developing-terragrunt, because they don't mention linters at all, but again i better describe all issues i encountered in separate issue or something.

yhakbar commented:
> It seems like docs are a bit outdated https://terragrunt.gruntwork.io/docs/community/contributing/#developing-terragrunt, because they don't mention linters at all, but again i better describe all issues i encountered in separate issue or something.

Please consider submitting a pull request to update them! We have a blind spot in that we develop Terragrunt day-in and day-out, so we rely on the community to propose adjustments to the docs when they aren't clear for new contributors.
